### ISSUE TITLE
No longer possible to request code coverage output path without running tests

### ISSUE THREAD

**Issue by gwynne** [apple/swift-package-manager#5853]
### Description

Ever since #5670 was merged, it is no longer possible to request the coverage output path without running tests in the process, even if they have already been run and a coverage report has already been generated. This significantly complicates building reusable CI logic where gathering code coverage may or may not be desirable for any given test run, as well as compromising the direct visibility of `swift test`'s overall output (with the new semantics of the `--show-codecov-path` option, the entire output of the test run must be captured in order to obtain the path).

It is not at all clear why code coverage must already have been generated in order to know the expected output path for the report - the PR's motivation says that requesting the path is only relevant once coverage has been gathered - but to my knowledge, the path is a matter of package configuration (and/or commandline flags), and is not affected by the actual gathering of coverage data at all.

Finally, there is no similar limitation on `swift build --show-bin-path`, which one could as easily say isn't relevant until a build has been completed. The change in behavior for one type of path but not another feels inconsistent; I don't understand what the benefit is intended to be. I'd much prefer for the behavior to be reverted, or at least to have a better explanation of why the change was made.

### Expected behavior

`swift test --show-codecov-path` should output the expected output location of a code coverage report without running the tests.

### Actual behavior

`swift test --show-codecov-path` now requires `--enable-code-coverage`, and `swift test --enable-code-coverage --show-codecov-path` causes the tests to run and coverage information to be gathered before outputting the path.

### Steps to reproduce

N/A

### Swift Package Manager version/commit hash

97503a15141e2de70853425c07677f9468ab99d8

### Swift & OS version (output of `swift --version && uname -a`)

```
root@12ec7f931acd:/# swift --version
uSwift version 5.8-dev (LLVM 170e2cfa496ad7c, Swift f32ae6f8c6068cd)
Target: aarch64-unknown-linux-gnu
root@12ec7f931acd:/# uname -a
Linux 12ec7f931acd 5.15.49-linuxkit #1 SMP PREEMPT Tue Sep 13 07:51:32 UTC 2022 aarch64 aarch64 aarch64 GNU/Linux
```

**Comment by neonichu** [comment]
cc @tomerd

**Comment by gwynne** [comment]
ping @neonichu @tomerd

**Comment by tomerd** [comment]
thanks for bringing this up @gwynne. the previous code had computed the path separately from the actual invocation and assumed the argument aligned which is very fragile.

what is the use case for acquiring the path without running the coverage? can you elaborate more on how you are using this? seems like there is some specific workflow that sounds a bit fragile

**Comment by gwynne** [comment]
In Vapor's case, we acquire the path as part of a separate GitHub Action used specifically as a wrapper around CodeCov's action in order to automate the conversion of coverage data to the LCOV format; the point of usage is here: https://github.com/vapor/swift-codecov-action/blob/main/action.yml#L43

The action is designed to be a drop-in replacement for the [CodeCov upload action](https://github.com/codecov/codecov-action) it wraps (notwithstanding that I never got around to making the input parameter names line up better); it has no control over the original `swift test` invocation or access to the output of that command (nor is it especially trivial to make GitHub Actions extract and pass along only the one relevant line of that output). In addition, the action _ought_ to be able to be invoked in the "aggregate" (e.g. processing multiple code coverage outputs in one go, again in parity with CodeCov), a mode of operation all but forbidden by this behavior change.

In short, I think it's reasonable for an action whose intent is already specific to SwiftPM packages to only have to be told the package to operate on instead of leaving the onus of recording the coverage file path (and/or preserving the file as needed) on the calling workflow - the entire point is to _minimize_ the amount of repetition that has to show up in the CI of every package that wants to have code coverage (and reusable workflows aren't much of a counterargument in their present limited form, unfortunately).

**Comment by neonichu** [comment]
Seems similar to `--show-bin-path` to me and we should be able to know these paths with a reasonable level of confidence without having to perform the action, right?

**Comment by gwynne** [comment]
@tomerd Any chance of getting this backported into 5.8?

**Comment by michalszelagsonos** [comment]
I am curious if this is also related to another issue where building tests separately and then running tests with coverage leads to an error:
```
error: The folder “codecov” doesn’t exist
```
Steps to reproduce, last command runs tests but fails at the end:
```
swift package clean
swift build --build-tests
swift test --skip-build --enable-code-coverage
```
Why would I want to do this?  My CI pipeline has instrumentation around how long steps take so I can keep tabs on how long our builds and tests take to execute.

**Comment by gwynne** [comment]
My guess (and I have to emphasize the word "guess" here) would be that `--skip-build` and `--enable-code-coverage` don't play well together, which is a similar but not really directly related issue.

**Comment by gwynne** [comment]
Closing this since it is indeed fixed in 5.9; it's unfortunate that it's still broken in 5.8, though.
